pkg/template/controller/templateinstance_controller.go

```diff
@@ -1,103 +1,202 @@
-	restconfig     rest.Config
+	oc             *client.Client
+	kc             kclientsetinternal.Interface
 	templateclient internalversiontemplate.TemplateInterface
-	controller     cache.Controller
+
+	lister   templatelister.TemplateInstanceLister
+	informer cache.SharedIndexInformer
+
+	queue workqueue.RateLimitingInterface
 }
 
-func NewTemplateInstanceController(restconfig rest.Config) *TemplateInstanceController {
-	c := TemplateInstanceController{restconfig: restconfig, templateclient: templateclientset.NewForConfigOrDie(&restconfig).Template()}
-	_, c.controller = cache.NewInformer(
-		&cache.ListWatch{
-			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-				return c.templateclient.TemplateInstances(kapi.NamespaceAll).List(options)
-			},
-			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
-				return c.templateclient.TemplateInstances(kapi.NamespaceAll).Watch(options)
-			},
+func NewTemplateInstanceController(oc *client.Client, kc kclientsetinternal.Interface, templateclient internalversiontemplate.TemplateInterface, informer internalversion.TemplateInstanceInformer) *TemplateInstanceController {
+	c := &TemplateInstanceController{
+		oc:             oc,
+		kc:             kc,
+		templateclient: templateclient,
+		lister:         informer.Lister(),
+		informer:       informer.Informer(),
+		queue:          workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "TemplateInstanceController"),
+	}
+
+	c.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueue(obj.(*templateapi.TemplateInstance))
 		},
-		&templateapi.TemplateInstance{},
-		0,
-		cache.ResourceEventHandlerFuncs{
-			AddFunc: func(obj interface{}) {
-				c.handle(obj.(*templateapi.TemplateInstance))
-			},
-			UpdateFunc: func(_, obj interface{}) {
-				c.handle(obj.(*templateapi.TemplateInstance))
-			},
-			DeleteFunc: func(obj interface{}) {
-			},
+		UpdateFunc: func(_, obj interface{}) {
+			c.enqueue(obj.(*templateapi.TemplateInstance))
 		},
-	)
+		DeleteFunc: func(obj interface{}) {
+		},
+	})
 
-	return &c
+	return c
 }
 
-func (c *TemplateInstanceController) Run(stop <-chan struct{}) {
-	c.controller.Run(stop)
+func (c *TemplateInstanceController) getTemplateInstanceCopy(key string) (*templateapi.TemplateInstance, error) {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	templateInstance, err := c.lister.TemplateInstances(namespace).Get(name)
+	if err != nil {
+		return nil, err
+	}
+
+	templateInstanceCopy, err := kapi.Scheme.DeepCopy(templateInstance)
+	if err != nil {
+		return nil, err
+	}
+
+	return templateInstanceCopy.(*templateapi.TemplateInstance), nil
 }
 
-func (c *TemplateInstanceController) handle(templateInstance *templateapi.TemplateInstance) error {
+func (c *TemplateInstanceController) sync(key string) error {
+	templateInstance, err := c.getTemplateInstanceCopy(key)
```

Comment:
you don't normally make a copy unless you're very likely to make a change to the object.  Looks like you have a quick check a few lines down which should be made before the copy is made.